client/rest/src/main/java/org/elasticsearch/client/Node.java

```diff
@@ -0,0 +1,235 @@
+        this.host = host;
+        this.boundHosts = boundHosts;
+        this.name = name;
+        this.version = version;
+        this.roles = roles;
+    }
+
+    /**
+     * Create a {@linkplain Node} without any metadata.
+     */
+    public Node(HttpHost host) {
+        this(host, null, null, null, null);
+    }
+
+    /**
+     * Make a copy of this {@link Node} but replacing its
+     * {@link #getHost() host}. Use this when the sniffing implementation
+     * returns a {@link #getHost() host} that is not useful to the client.
+     */
+    public Node withHost(HttpHost host) {
+        /*
+         * If the new host isn't in the bound hosts list we add it so the
+         * result looks sane.
+         */
+        Set<HttpHost> boundHosts = this.boundHosts;
+        if (false == boundHosts.contains(host)) {
+            boundHosts = new HashSet<>(boundHosts);
+            boundHosts.add(host);
+            boundHosts = unmodifiableSet(boundHosts);
+        }
+        return new Node(host, boundHosts, name, version, roles);
+    }
+
+    /**
+     * Contact information for the host.
+     */
+    public HttpHost getHost() {
+        return host;
+    }
+
+    /**
+     * Addresses on which the host is listening. These are useful to have
+     * around because they allow you to find a host based on any address it
+     * is listening on.
+     */
+    public Set<HttpHost> getBoundHosts() {
+        return boundHosts;
+    }
+
+    /**
+     * The {@code node.name} of the node.
+     */
+    public String getName() {
+        return name;
+    }
+
+    /**
+     * Version of Elasticsearch that the node is running or {@code null}
+     * if we don't know the version.
+     */
+    public String getVersion() {
+        return version;
+    }
+
+    /**
+     * Roles that the Elasticsearch process on the host has or {@code null}
+     * if we don't know what roles the node has.
+     */
+    public Roles getRoles() {
+        return roles;
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder b = new StringBuilder();
+        b.append("[host=").append(host);
+        if (boundHosts != null) {
+            b.append(", bound=").append(boundHosts);
+        }
+        if (name != null) {
+            b.append(", name=").append(name);
+        }
+        if (version != null) {
+            b.append(", version=").append(version);
+        }
+        if (roles != null) {
+            b.append(", roles=").append(roles);
+        }
+        return b.append(']').toString();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == null || obj.getClass() != getClass()) {
+            return false;
+        }
+        Node other = (Node) obj;
+        return host.equals(other.host)
+            && Objects.equals(boundHosts, other.boundHosts)
+            && Objects.equals(version, other.version)
+            && Objects.equals(name, other.name)
+            && Objects.equals(roles, other.roles);
```

Comment:
this may sound controversial but in my mind the answer to the "when are two nodes equal?" question is "when they have the same host and port".

Reply:
shall we then drop these methods in this case, I think it confuses me to have equals working this way if most of the times we want to compare the hosts :)